### PR TITLE
Shelan's API Muncher

### PR DESCRIPTION
**Pull request by sheland**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | I utilized Postman where I looked at the JSON returned after inputting required parameters, app ID and Key to view what a recipe looks like. JSON for recipes is a giant array with hashes and arrays that contains recipe fields  
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | An API wrapper lets you call the API instead of the API directly. If you want to change from one API to another or if changes occur in the API, instead of rewriting your code, all you have to change is the wrapper. It it located in the 'lib' because rails doesn't automatically know about custom libraries lib folder. It does not have any magic around consuming APIs like it does for models or views. 

Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful | I created 2 methods that list recipes and shows a recipe. I created 2 more methods ".create_recipe" and "self.keyboard" that creates a recipes and a recipe. 
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? | I did not test edge or failure cases just nominal cases "it can list recipes' and "it can show a recipe"
How does VCR aid in testing an API? | It prevents you from calling the API by recording HTTP interactions and replay them . 
What is the Heroku URL of your deployed application? | https://ada-muncher-2018.herokuapp.com/

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'recipe#home'
+
+  get '/recipes', to: 'recipe#index', as: 'recipes'
+  get '/recipe/:uri', to: 'recipe#show', as: 'recipe'
```
Comment (dHelmgren) [inline config/routes.rb:7]: Could you use `resources` for the index and show pages?

File: lib/edamam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      image: api_params["recipe"]["image"],
+      serving: api_params["recipe"]["yield"],
+      dietLabels: api_params["recipe"]["dietLabels"],
+      healthLabels: api_params["recipe"]["healthLabels"],
+      ingredients: api_params["recipe"]["ingredients"],
+      calories: api_params["recipe"]["calories"]
+    )
+  end
+
+  def self.keyboard(api_params)
```
Comment (dHelmgren) [inline lib/edamam_api_wrapper.rb:53]: The `keyboard` method here seems like the thing you want to call when you do `create_recipe` as well. This code isn't very DRY.

File: app/controllers/recipe_controller.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+
+class RecipeController < ApplicationController
+  def index
+    @recipes = EdamamApiWrapper.search(params[:food]).paginate(page: params[:page], per_page: 10)
+    # @recipes = EdamamApiWrapper.search(params[:food]).paginate page: params[:page], per_page: 10
+  end
+
+  def show
+    uri = params[:uri]
+    @recipe = EdamamApiWrapper.find_a_recipe(uri).first #takes the object out []
```
Comment (dHelmgren) [inline app/controllers/recipe_controller.rb:10]: You've written your API wrapper to raise an error if the API call fails, but you're not looking for an error here. You should wrap this call in a `begin`/`rescue`.

File: test/controllers/recipe_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+require "test_helper"
+
+describe RecipeController do
+  it "should get index" do
+    VCR.use_cassette('search_file') do
+      get recipes_path
+       must_respond_with :success
```
Comment (dHelmgren) [inline test/controllers/recipe_controller_test.rb:7]: There are some interesting cases you're not covering here:

- What happens if no search term is provided?
- What if the search returns no results?

It might also be worthwhile to add some tests around the paging parameters:

- What happens if they're absent?
- Do you get different results when they change?
- What if you send a bogus value, like a negative page number?

File: test/lib/edamam_api_wrapper_test.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
+
+describe EdamamApiWrapper do
+
+  it 'can list recipes' do
+    VCR.use_cassette('seach_file') do  #ensure code inside block will use cassette
+      hits = EdamamApiWrapper.search('food')
```
Comment (dHelmgren) [inline test/lib/edamam_api_wrapper_test.rb:7]: What if the search term has no results?

File: test/lib/edamam_api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        expect(hit).must_respond_to :uri
+        expect(hit).must_respond_to :label
+      end
+    end
+  end
+
+  it 'can show a recipe'do
+    VCR.use_cassette('find_a_recipe') do
+      apple_pie = EdamamApiWrapper.find_a_recipe('pie')
+
```
Comment (dHelmgren) [inline test/lib/edamam_api_wrapper_test.rb:20]: What if the recipe doesn't exist?

**Comment by dHelmgren** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene | Good, but more frequent commits are a good idea.
Comprehension questions | You missed a piece on question 2! Otherwise good.
**General** |
Rails fundamentals (RESTful routing, use of named paths) | Good
Semantic HTML | Yes
Errors are reported to the user | Not well. See comments in the controller
API Wrapper to handle the API requests | yes
Controller testing | A little too basic. see comments
Lib testing | See comments
Search Functionality | yes
List Functionality | yes, but see comments
Show individual item functionality | yes, see comments
**Styling** |
List view shows 10 items at a time and/or has pagination | yes
Attractive and usable UI | Simple, functional. Your page counters ended up in a weird spot though.
**API Features** |
The App attributes Edamam | main page does, but the other pages don't. This is really important! When you're designing code for money, using someone else's work without attribution is a good way to end up in legal trouble.
The VCR cassettes do not contain the API key | They, don't, good work!
**External Resources** |
Link to deployed app on Heroku | yes
**Overall** | Good work, you covered the majority of the learning goals for this assignment. The most important thing that needs your attention is really the lack of attribution to Edamam, as that could lose you lots of money in the real world. Besides that, I REALLY want to encourage you to engage in test driven development as a part of this process: it takes more time up front, but when we start to run into issues as developers, knowing where the problems are from the tests saves a lot of time.
